include/pagmo/problems/cec2013.hpp

```diff
@@ -0,0 +1,1218 @@
+ *
+ * @see http://www.ntu.edu.sg/home/EPNSugan/index_files/CEC2013/CEC2013.htm
+ * @see http://web.mysites.ntu.edu.sg/epnsugan/PublicSite/Shared%20Documents/CEC2013/cec13-c-code.zip
+ */
+class cec2013
+{
+public:
+    /// Default Constructor
+    /**
+     * Will construct an empty object
+     *
+     * **NOTE** This constructor should never be used and is here only for serialization
+     */
+    cec2013() : m_prob_id(1), m_rotation_matrix(), m_origin_shift(), m_y(2), m_z(2) {}
+    /// Constructor
+    /**
+     * Will construct one of the 28 CEC2013 problems
+     *
+     * @param[in] prob_id The problem id. One of [1,2,...,28]
+     * @param[in] dim problem dimension. One of [2,5,10,20,30,...,100]
+     * @param[in] data_dir The path where the CEC2013 data files are located.
+     *
+     * **NOTE** Two files are expected to be in \p dir "M_Dx.txt" and "shift_data.txt", where "x" is
+     * the problem dimension \p dim
+     *
+     * @see http://web.mysites.ntu.edu.sg/epnsugan/PublicSite/Shared%20Documents/CEC2013/cec13-c-code.zip to find
+     * the files
+     *
+     * @throws invalid_argument if \p prob_id is not in [1,18] or if \p dim is not one of
+     * [2,5,10,20,30,40,50,60,70,80,90,100]
+     * @throws io_error if the files are not found
+     */
+    cec2013(unsigned int prob_id, unsigned int dim, const std::string &data_dir = "input_data/")
+        : m_prob_id(prob_id), m_y(dim), m_z(dim)
+    {
+        if (!(dim == 2u || dim == 5u || dim == 10u || dim == 20u || dim == 30u || dim == 40u || dim == 50u || dim == 60u
+              || dim == 70u || dim == 80u || dim == 90u || dim == 100u)) {
+            pagmo_throw(std::invalid_argument, "Error: CEC2013 Test functions are only defined for dimensions "
+                                               "2,5,10,20,30,40,50,60,70,80,90,100, a dimension of "
+                                                   + std::to_string(dim) + " was detected.");
+        }
+        if (prob_id < 1u || prob_id > 18u) {
+            pagmo_throw(std::invalid_argument,
+                        "Error: CEC2013 Test functions are only defined for prob_id in [1, 18], a prob_id of "
+                            + std::to_string(prob_id) + " was detected.");
+        }
+
+        std::string data_file_name(data_dir);
+        // We create the full file name for the shift vector
+        data_file_name.append("shift_data.txt");
+        // And we read all data into m_origin_shift
+        {
+            std::ifstream data_file(data_file_name.c_str());
+            if (!data_file.is_open()) {
+                pagmo_throw(std::ios_base::failure,
+                            std::string("Error: file not found. I was looking for ").append(data_file_name.c_str()));
+            }
+            std::istream_iterator<double> start(data_file), end;
+            m_origin_shift = std::vector<double>(start, end);
+            data_file.close();
```

Comment:
The file destructors close the file, so it is not necessary to call explicitly the ``close()`` method here.